pkg/ocmagenthandler/ocmagenthandler.go

```diff
@@ -0,0 +1,101 @@
+package ocmagenthandler
+
+import (
+	"context"
+	"github.com/go-logr/logr"
+	ocmagentv1alpha1 "github.com/openshift/ocm-agent-operator/pkg/apis/ocmagent/v1alpha1"
+	"k8s.io/apimachinery/pkg/runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+//go:generate mockgen -source $GOFILE -destination ../../pkg/util/test/generated/mocks/$GOPACKAGE/interfaces.go -package $GOPACKAGE
+
+type OCMAgentHandler interface {
+	// EnsureOCMAgentResourcesExist ensures that an OCM Agent is deployed on the cluster.
+	EnsureOCMAgentResourcesExist(ocmagentv1alpha1.OcmAgent) error
+	// EnsureOCMAgentResourcesAbsent ensures that all OCM Agent resources are removed on the cluster.
+	EnsureOCMAgentResourcesAbsent(ocmagentv1alpha1.OcmAgent) error
+}
+
+type ocmAgentHandler struct {
+	Client client.Client
+	Scheme *runtime.Scheme
+	Log    logr.Logger
+	Ctx    context.Context
+}
+
+func New(client client.Client, scheme *runtime.Scheme, log logr.Logger, ctx context.Context) OCMAgentHandler {
+	return &ocmAgentHandler{
+		Client: client,
+		Scheme: scheme,
+		Log:    log,
+		Ctx:    ctx,
+	}
+}
+
+func (o *ocmAgentHandler) EnsureOCMAgentResourcesExist(ocmAgent ocmagentv1alpha1.OcmAgent) error {
+	// Ensure we have a Deployment
+	o.Log.V(2).Info("Entering ensureDeployment")
+	err := o.ensureDeployment(ocmAgent)
+	if err != nil {
+		return err
+	}
+
+	// Ensure we have a ConfigMap
+	o.Log.V(2).Info("Entering ensureConfigMap")
+	err = o.ensureConfigMap(ocmAgent)
+	if err != nil {
+		return err
+	}
+
+	// Ensure we have a Secret
+	o.Log.V(2).Info("Entering ensureAccessTokenSecret")
+	err = o.ensureAccessTokenSecret(ocmAgent)
+	if err != nil {
+		return err
+	}
+
+	// Ensure we have a service
+	o.Log.V(2).Info("Entering ensureService")
+	err = o.ensureService(ocmAgent)
+	if err != nil {
+		return err
+	}
+
+	// TODO: Ensure we have a ServiceMonitor
+
+	return nil
+}
+
+func (o *ocmAgentHandler) EnsureOCMAgentResourcesAbsent(ocmAgent ocmagentv1alpha1.OcmAgent) error {
```

Comment:
same here, can we register the steps in somewhere and run them in a loop?

Reply:
I'll have a think about this and see if it can be improved some. I'm a little bit wary about over-engineering it simply for the sake of doing so - does it buy us that much in improving readability or complexity?

Reply:
Just thought they are all doing the same thing, running in a loop may make it look more clean, and should also make the file shorter? I don't insist on this.
Seems they are calling different functions, so it will not be as easy as we put them into an array. and just call them in a for loop?

Reply:
I've pushed an update as part of the latest commit, is this what you were thinking? Direct link to the reworked function [here](https://github.com/openshift/ocm-agent-operator/blob/c534bc1402667abf522394b31d844892f7e5fdfa/pkg/ocmagenthandler/ocmagenthandler.go#L38).

Reply:
ah, cool, yes, that's what I am thinking. Thanks!